catroid/src/main/java/org/catrobat/catroid/embroidery/EmbroideryListManager.java

```diff
@@ -0,0 +1,85 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.catrobat.catroid.embroidery;
+
+import android.graphics.PointF;
+
+import org.catrobat.catroid.content.Sprite;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.TreeMap;
+
+public class EmbroideryListManager {
+	private Map<String, Integer> currentSpriteLayerMap = new HashMap<>();
+	private Map<Integer, Map<String, EmbroideryList>> layerSpriteEmbroideryListMap = new TreeMap<>();
+
+	public void reset() {
+		layerSpriteEmbroideryListMap = new TreeMap<>();
+		currentSpriteLayerMap = new HashMap<>();
+	}
+
+	public synchronized boolean addEmbroideryPointOfSprite(PointF point, Sprite sprite) {
+		if (!layerSpriteEmbroideryListMap.containsKey(sprite.look.getZIndex())) {
+			layerSpriteEmbroideryListMap.put(sprite.look.getZIndex(), new TreeMap<>());
+			layerSpriteEmbroideryListMap.get(sprite.look.getZIndex()).put(sprite.getName(), new EmbroideryList());
+		} else if (!layerSpriteEmbroideryListMap.get(sprite.look.getZIndex()).containsKey(sprite.getName())) {
+			layerSpriteEmbroideryListMap.get(sprite.look.getZIndex()).put(sprite.getName(), new EmbroideryList());
+		}
+
+		if (currentSpriteLayerMap.containsKey(sprite.getName()) && currentSpriteLayerMap.get(sprite.getName()) != sprite.look.getZIndex()) {
+			EmbroideryList spriteEmbroideryList =
+					layerSpriteEmbroideryListMap.get(currentSpriteLayerMap.get(sprite.getName())).get(sprite.getName());
+			StitchPoint stitchPoint = (StitchPoint) spriteEmbroideryList.get(spriteEmbroideryList.size() - 1);
+			layerSpriteEmbroideryListMap.get(sprite.look.getZIndex()).get(sprite.getName()).add(stitchPoint.point);
+			currentSpriteLayerMap.put(sprite.getName(), sprite.look.getZIndex());
+		}
+
+		currentSpriteLayerMap.put(sprite.getName(), sprite.look.getZIndex());
+		return layerSpriteEmbroideryListMap.get(sprite.look.getZIndex()).get(sprite.getName()).add(point);
+	}
+
+	public EmbroideryList getEmbroideryList() {
+		EmbroideryList embroideryList = new EmbroideryList();
+
+		for (Map.Entry<Integer, Map<String, EmbroideryList>> layer : layerSpriteEmbroideryListMap.entrySet()) {
+			for (Map.Entry<String, EmbroideryList> sprite : layer.getValue().entrySet()) {
+
+				EmbroideryList list = sprite.getValue();
+
+				embroideryList.addAll(list);
+				StitchPoint lastStitchOfSprite = (StitchPoint) list.get(list.size() - 1);
+				embroideryList.add(new StitchPoint(lastStitchOfSprite.point, false, true));
+				embroideryList.add(new StitchPoint(lastStitchOfSprite.point, true, false));
+			}
+		}
+
+		if (!embroideryList.isEmpty()) {
+			embroideryList.remove(embroideryList.size() - 1);
```

Comment:
this is going to break as soon as you have 1 element in it due to out of bounds exception